### ISSUE TITLE
External endpoints setup for failover using ServiceEntry locality feature must support DNS resolution

### ISSUE THREAD

**Issue by uk000** [istio/istio#14496]
**Bug description**
As of Istio 1.1.7, ServiceEntry locality feature (for adding additional external endpoints for a service) only supports STATIC resolution and requires endpoint destinations specified as IP addresses. In any non-trivial production-grade scenario where a service is running on cloud VMs as well as inside a kubernetes cluster, and the service wants to enable in-cluster traffic to failover to its external endpoints, requiring the configuration to use IP addresses only is impractical. Service VMs can and will go up and down, and the service's endpoint IP addresses will change frequently. It's impractical to expect the service owner to keep the ServiceEntry up-to-date whenever the service's external endpoints change.

**Expected behavior**
The locality based routing feature of ServiceEntry needs to support DNS resolution, so that a service can specify its external endpoint as an FQDN that gets resolved to its available IP addresses automatically.

**Steps to reproduce the bug**
Follow the same steps as in issue #14495 

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
```
version.BuildInfo{Version:"1.1.7", GitRevision:"eec7a74473deee98cad0a996f41a32a47dd453c2", User:"root", Host:"341b3bf0-76ac-11e9-b644-0a580a2c0404", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.6-6-geec7a74"}
```
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:11:31Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-08T17:02:58Z", GoVersion:"go1.12.1", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
Helm template

**Environment where bug was observed (cloud vendor, OS, etc)**
Docker for mac

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by howardjohn** [comment]
@PiotrSikora any insight on how feasible this is? Our locality works by setting the [priority](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/load_balancing/priority) field. So in this case they have endpoints in cluster with priority 0 and endpoints external endpoints with priority 1.

**Comment by howardjohn** [comment]
Also a minor clarification - locality lb does work with DNS resolution, but dns resolution won't merge with an existing service so you can't failover from kubernetes service to external service

**Comment by rshriram** [comment]
Yes. this is more of an envoy restriction

**Comment by howardjohn** [comment]
Assigned to Yan who is working on this on envoy side. I'll pick up the work to configure it in Pilot side once it is ready

**Comment by crazyxy** [comment]
I am not sure if I understand the requirement correctly. Do we want to support external service with FQDN for locality lb?

I added an external service using FQDN to the the locality endpoints. The config dump is like:

```json
"cluster": {
    "name": "server",
    "type": "STRICT_DNS",
    "connect_timeout": "0.250s",
    "load_assignment": {
        "cluster_name": "server",
        "endpoints": [
            {
                "locality": {
                    "region": "US-West",
                    "zone": "az1",
                    "sub_zone": "sub1"
                },
                "lb_endpoints": [
                    {
                        "endpoint": {
                            "address": {
                                "socket_address": {
                                    "address": "server",
                                    "port_value": 80
                                }
                            }
                        }
                    }
                ]
            },
            {
                "locality": {
                    "region": "Asia-East",
                    "zone": "az1",
                    "sub_zone": "sub1"
                },
                "lb_endpoints": [
                    {
                        "endpoint": {
                            "address": {
                                "socket_address": {
                                    "address": "httpbin.org",
                                    "port_value": 80
                                }
                            }
                        }
                    },
                    {
                        "endpoint": {
                            "address": {
                                "socket_address": {
                                    "address": "server",
                                    "port_value": 443
                                }
                            }
                        }
                    }
                ]
            }
        ]
    }
}
```

I found that the client can access `httpbin.org` and the `/cluster` shows healthy upstream.

```
server::52.0.57.170:80::cx_active::5
server::52.0.57.170:80::cx_connect_fail::0
server::52.0.57.170:80::cx_total::8
server::52.0.57.170:80::rq_active::0
server::52.0.57.170:80::rq_error::0
server::52.0.57.170:80::rq_success::17
server::52.0.57.170:80::rq_timeout::0
server::52.0.57.170:80::rq_total::20
server::52.0.57.170:80::hostname::httpbin.org
server::52.0.57.170:80::health_flags::healthy
server::52.0.57.170:80::weight::1
server::52.0.57.170:80::region::Asia-East
server::52.0.57.170:80::zone::az1
server::52.0.57.170:80::sub_zone::sub1
server::52.0.57.170:80::canary::false
server::52.0.57.170:80::priority::0
server::52.0.57.170:80::success_rate::-1
server::52.0.57.170:80::local_origin_success_rate::-1
server::34.230.136.58:80::cx_active::5
server::34.230.136.58:80::cx_connect_fail::0
server::34.230.136.58:80::cx_total::9
server::34.230.136.58:80::rq_active::0
server::34.230.136.58:80::rq_error::0
server::34.230.136.58:80::rq_success::12
server::34.230.136.58:80::rq_timeout::0
server::34.230.136.58:80::rq_total::16
server::34.230.136.58:80::hostname::httpbin.org
server::34.230.136.58:80::health_flags::healthy
server::34.230.136.58:80::weight::1
server::34.230.136.58:80::region::Asia-East
server::34.230.136.58:80::zone::az1
server::34.230.136.58:80::sub_zone::sub1
server::34.230.136.58:80::canary::false
server::34.230.136.58:80::priority::0
server::34.230.136.58:80::success_rate::-1
server::34.230.136.58:80::local_origin_success_rate::-1
```
`34.230.136.58` and `52.0.57.170` are two IPs resolved by DNS for `httpbin.org`.

**Comment by howardjohn** [comment]
In your case you have cluster type `STRICT_DNS` which works fine as you mentioned.

The problem is that for kubernetes services, Pilot will use cluster type EDS.

The use case is we should be able to have request sent in cluster to kubernetes services, and failover to external service. You can do this today by adding a ServiceEntry with resolution type STATIC and specifying an IP address. Pilot will then add this IP to the set of endpoints in the EDS response.

This works but there are still some problems. 

* It isn't feasible to add an ip address. I would want to say failover to `httpbin.org` not `1.2.3.4`
* Cluster settings may be different for external service and internal -- specifically, mtls is problematic

Essentially what we need is to have some sort of "hybrid cluster".

Right now this can be worked around with an egressgateway: Add the IP address of the gateway (should be pretty static), and then have it failover to the gateway. This is a pretty hacky solution though: we would be adding a completely separate Envoy just to get this "hybrid cluster" we want.

**Comment by crazyxy** [comment]
Issue in Envoy https://github.com/envoyproxy/envoy/issues/7454

**Comment by rshriram** [comment]
I thikn the issue is simpler. We simply need to support the ability for a service entry to override the inferred loadbalancing type of a cluster. If you specify a service entry such as 

```
host: productpage.ns1.svc.cluster.local
resolution: DNS
endpoints:
 - httpbin.org
   locality: foo
```
and k8s has a productpage service with 5 pods in locality bar, the CDS cluster emitted by pilot for Envoy should be of the form:
```
v2.Cluster
lbType: STRICT_DNS
Hosts:
 1.1.1.1 (locality bar)
 2.2.2.2 (locality bar)
 3.3.3.3 (locality bar)
 httpbin.org (locality foo)
```

With the locality setting, you can achieve the failover property you want. The thing that needs to be doable in Pilot is that we need to be able to merge the endpoints from service entries and k8s (across lb types) as well as overwrite the service's resolution type (from default client-side-lb to dns resolution). 

This should have been easily doable except, you folks also wanted to have multiple service entries for the same host in different namespaces, and have the service entry be interpreted in a namespace specific manner. This semantic implies override (not merge), where the service in the local or imported namespace wins over something else someone has declared. @howardjohn just implemented it after a few weeks of iteration. :(.

So currently, your only option is to simply declare a service entry of the following form:
```
Host: productpage.ns1.svc.cluster.local
Resolution: DNS
Endpoints:
- productpage.ns1.svc.cluster.local, locality foo
- myvm.org, locality bar
```

When you set this, we can replace the usual EDS cluster with the strictDNS cluster. envoy will send the traffic by default to internal k8s service (productpage.ns1.svc.cluster.local), based on locality weights, and failover to myvm.org.

I know that you also want to do mTLS inside the cluster for k8s pods, but that is a separate issue being tackled. The failover piece should be addressable using the above technique due to the limitations imposed by your former requirement (same service in different namespaces, with the closest one wins).

**Comment by howardjohn** [comment]
@rshriram if we do
```
v2.Cluster
lbType: STRICT_DNS
Hosts:
 1.1.1.1 (locality bar)
 2.2.2.2 (locality bar)
 3.3.3.3 (locality bar)
 httpbin.org (locality foo)
```

 Does this work? Will envoy correctly handle an IP address and not try to do a DNS lokup? We certainly could not do incremental EDS updates if we do this

**Comment by rshriram** [comment]
Yes, envoy handles both IPs and dns names in the strict dns cluster.

**Comment by uk000** [comment]
@rshriram "overwrite the service's resolution type" I guess that's the key and it should work. What I didn't understand is why's this in conflict with having each namespace apply its own service entry. Doesn't Pilot resolve the clusters to be sent for each envoy individually? Then wouldn't Pilot be able to figure out that for ns A, it needs to add the "external" endpoints that it read from SE from nsA, and similar for nsB. So even if ns A and B have two ServiceEntry defining the same fqdn with different endpoint resolutions, they won't step on each other.

**Comment by costinm** [comment]
Using DNS means most load blanacing capabilities are lost - and we rely on DNS to handle load balancing. I don't think this is ideal. 

While it's true that currently Pilot is not supporting health/load reports and it is not actually doing advanced subseting - we are planning to do it, and any solution based on not using EDS would be
incompatible with more advanced load balancing.

**Comment by costinm** [comment]
An opposite solution would be for Pilot to resolve the DNS fallback address, and treat the result as discovery info pushed in EDS. The problem is that more advanced DNS load balancers would be confused - but it might work well with simple round-robin DNS.

**Comment by tangmengqiu** [comment]
@howardjohn any update for this? I have the same requirements.
